File: src/main/java/com/github/samarium150/structurescompass/gui/StructureSearchEntry.java
package com.github.samarium150.structurescompass.gui;

import com.github.samarium150.structurescompass.util.StructureUtils;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import net.minecraft.Util;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.components.ObjectSelectionList.Entry;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.client.resources.sounds.SimpleSoundInstance;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.TextComponent;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.world.level.levelgen.feature.StructureFeature;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;

import javax.annotation.Nonnull;

/**
 * Class of entries in the list for searching
 * <p>
 * This class is adapted from
 * <a href="https://github.com/MattCzyr/NaturesCompass" target="_blank">NaturesCompass</a>
 * which is under
 * <a href="https://creativecommons.org/licenses/by-nc-sa/4.0" target="_blank">
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License
 * </a>.
 */
@OnlyIn(Dist.CLIENT)
public final class StructureSearchEntry extends Entry<StructureSearchEntry> {
    
    private final Minecraft minecraft;
    private final StructureSearchList list;
    private final StructureFeature<?> structure;
    private final StructuresCompassScreen screen;
    private long lastTickTime;
    
    public StructureSearchEntry(@Nonnull StructureSearchList list, StructureFeature<?> structure) {
        minecraft = Minecraft.getInstance();
        this.list = list;
        this.structure = structure;
        screen = list.getScreen();
    }
    
    public void search() {
        minecraft.getSoundManager().play(SimpleSoundInstance.forUI(SoundEvents.UI_BUTTON_CLICK, 1.0F));
        screen.search(structure);
    }

    public StructureFeature<?> getStructure() {
        return structure;
    }
    
    @Override
    public void render(
        @Nonnull PoseStack matrixStack,
        int index, int top, int left, int width, int height, int mouseX, int mouseY,
        boolean isMouseOver, float partialTicks
    ) {
        minecraft.font.draw(matrixStack,
            StructureUtils.getLocalizedStructureName(structure),
            left + 1, top + 1, 0xffffff
        );
        minecraft.font.draw(matrixStack,
            new TextComponent(I18n.get("string.structurescompass.source") + ": "
                                        + StructureUtils.getStructureSource(structure)),
            left + 1, top + minecraft.font.lineHeight + 3, 0x808080
        );
        minecraft.font.draw(matrixStack,
            new TextComponent(I18n.get("string.structurescompass.dimension") + ": "
                                        + StructureUtils.getDimensions(structure)),
            left + 1, top + minecraft.font.lineHeight + 14, 0x808080
        );
        RenderSystem.setShaderColor(1.0F, 1.0F, 1.0F, 1.0F);
    }
    
    @Override
    public boolean mouseClicked(double mouseX, double mouseY, int button) {
        if (button == 0) {
            list.selectStructure(this);
            if (Util.getMillis() - lastTickTime < 250L) {
                search();
                return true;
            }
            lastTickTime = Util.getMillis();
        }
        return false;
    }
    
    @Nonnull
    @Override
    public Component getNarration() {
        return new TextComponent(I18n.get("string.structurescompass.source") + ": "
                                     + StructureUtils.getStructureSource(structure));
    }
}

File: src/main/kotlin/io/github/samarium150/minecraft/mod/structures_compass/client/gui/widget/StructureSearchEntry.kt
/*
 * Copyright (c) 2022 Samarium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/gpl-3.0.html>.
 */
package io.github.samarium150.minecraft.mod.structures_compass.client.gui.widget

import com.mojang.blaze3d.systems.RenderSystem
import io.github.samarium150.minecraft.mod.structures_compass.client.util.getLocalizedDimensionName
import io.github.samarium150.minecraft.mod.structures_compass.client.util.getLocalizedStructureName
import io.github.samarium150.minecraft.mod.structures_compass.util.*
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.minecraft.client.gui.widget.AlwaysSelectedEntryListWidget.Entry
import net.minecraft.client.sound.PositionedSoundInstance
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.sound.SoundEvents
import net.minecraft.text.LiteralText
import net.minecraft.text.Text
import net.minecraft.text.TranslatableText
import net.minecraft.util.Identifier
import net.minecraft.util.Util
import net.minecraft.world.gen.feature.StructureFeature

@Environment(EnvType.CLIENT)
class StructureSearchEntry(
    private val searchList: StructureSearchList,
    val structure: StructureFeature<*>
) : Entry<StructureSearchEntry>() {

    private val client = minecraftClient
    private val screen = searchList.screen
    private var lastClickTime = 0L

    fun search() {
        client.soundManager.play(PositionedSoundInstance.master(SoundEvents.UI_BUTTON_CLICK, 1.0f))
        screen.search(structure)
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        return if (button == 0) {
            searchList.selectStructure(this)
            return if (Util.getMeasuringTimeMs() - lastClickTime < 250L) {
                search()
                true
            } else {
                lastClickTime = Util.getMeasuringTimeMs()
                false
            }
        } else false
    }

    override fun render(
        matrices: MatrixStack?,
        index: Int,
        y: Int,
        x: Int,
        entryWidth: Int,
        entryHeight: Int,
        mouseX: Int,
        mouseY: Int,
        hovered: Boolean,
        tickDelta: Float
    ) {
        client.textRenderer.draw(
            matrices,
            LiteralText(structure.getIdentifier()?.getLocalizedStructureName() ?: ""),
            (x + 1).toFloat(),
            (y + 1).toFloat(),
            0xffffff
        )
        client.textRenderer.draw(
            matrices,
            TranslatableText("${prefix}source").append(": ${structure.getNamespace()}"),
            (x + 1).toFloat(),
            (y + 3 + client.textRenderer.fontHeight).toFloat(),
            0x808080
        )
        client.textRenderer.draw(
            matrices,
            TranslatableText("${prefix}dimension")
                .append(": ${structure.getDimensions().map(Identifier::getLocalizedDimensionName)}"),
            (x + 1).toFloat(),
            (y + 14 + client.textRenderer.fontHeight).toFloat(),
            0x808080
        )
        RenderSystem.setShaderColor(1.0f, 1.0f, 1.0f, 1.0f)
    }

    override fun getNarration(): Text {
        return LiteralText(structure.getIdentifier()!!.getLocalizedStructureName())
    }
}
